J2V8 native library not loaded, libj2v8_macosx_x86_64.dylib already loaded in another classloader

If you are submitting a **bug**, please include the following:

- [x] summary of problem

  - If you use one version of spotless (e.g. `3.15.0-SNAPSHOT`)
  - and you run a `spotlessApply` that uses an npm rule
  - then use a new version (e.g. `3.15.0`)
  - and you run a `spotlessApply` that uses an npm rule
  - then you'll get the stacktrace copy-pasted later in this bug.

The workaround is easy, just run `gradlew --stop` to kill the old daemon.  Better would be to catch the error and tell the user "You can fix this with 'gradlew --stop'".  Best would be to fix it.

Not a high priority issue. :)

- [x] gradle or maven version: `4.10.2`
- [x] spotless version: `3.15.0`
- [x] operating system and version: `macOS 10.13.6`
- [x] copy-paste your full Spotless configuration block(s), and a link to a public git repo that reproduces the problem if possible

[Public repo is here.](https://github.com/mytakedotorg/mytakedotorg/pull/212/commits/4ebdbc68ccc3cfa67af9bb9b26e28f5d69a9497f)

```gradle
spotless {
	format 'scss', {
		target 'src/main/styles/**/*.scss'
		prettier().config(['parser': 'postcss'])
	}
	format 'typescript', {
		target 'src/main/scripts/**'
		prettier().config(['parser': 'typescript'])
	}
}
```

- [x] copy-paste the full content of any console errors emitted by `gradlew spotless[Apply/Check] --stacktrace`

```
Caused by: java.lang.IllegalStateException: J2V8 native library not loaded
        at com.eclipsesource.v8.V8.checkNativeLibraryLoaded(V8.java:195)
        at com.eclipsesource.v8.V8.createV8Runtime(V8.java:149)
        at com.eclipsesource.v8.V8.createV8Runtime(V8.java:125)
        at com.eclipsesource.v8.NodeJS.createNodeJS(NodeJS.java:58)
        at com.eclipsesource.v8.NodeJS.createNodeJS(NodeJS.java:45)
        at com.diffplug.spotless.npm.Reflective.invokeStaticMethod(Reflective.java:56)
        at com.diffplug.spotless.npm.NodeJSWrapper.lambda$new$0(NodeJSWrapper.java:39)
        at com.diffplug.spotless.npm.ReflectiveObjectWrapper.<init>(ReflectiveObjectWrapper.java:34)
        at com.diffplug.spotless.npm.NodeJSWrapper.<init>(NodeJSWrapper.java:33)
        at com.diffplug.spotless.npm.NpmFormatterStepStateBase.nodeJSWrapper(NpmFormatterStepStateBase.java:95)
        at com.diffplug.spotless.npm.PrettierFormatterStep$State.createFormatterFunc(PrettierFormatterStep.java:68)
        at com.diffplug.spotless.FormatterStepImpl$Standard.format(FormatterStepImpl.java:76)
        at com.diffplug.spotless.FormatterStep$Strict.format(FormatterStep.java:76)
        at com.diffplug.spotless.Formatter.compute(Formatter.java:230)
        at com.diffplug.spotless.Formatter.applyToAndReturnResultIfDirty(Formatter.java:192)
        at com.diffplug.gradle.spotless.SpotlessTask.applyAnyChanged(SpotlessTask.java:230)
        at com.diffplug.gradle.spotless.SpotlessTask.performAction(SpotlessTask.java:193)
        at org.gradle.internal.reflect.JavaMethod.invoke(JavaMethod.java:73)
        at org.gradle.api.internal.project.taskfactory.IncrementalTaskAction.doExecute(IncrementalTaskAction.java:50)
        at org.gradle.api.internal.project.taskfactory.StandardTaskAction.execute(StandardTaskAction.java:39)
        at org.gradle.api.internal.project.taskfactory.StandardTaskAction.execute(StandardTaskAction.java:26)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$1.run(ExecuteActionsTaskExecuter.java:131)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:300)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:292)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:174)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:90)
        at org.gradle.internal.operations.DelegatingBuildOperationExecutor.run(DelegatingBuildOperationExecutor.java:31)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeAction(ExecuteActionsTaskExecuter.java:120)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeActions(ExecuteActionsTaskExecuter.java:99)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:77)
        at org.gradle.api.internal.tasks.execution.OutputDirectoryCreatingTaskExecuter.execute(OutputDirectoryCreatingTaskExecuter.java:51)
        at org.gradle.api.internal.tasks.execution.SkipUpToDateTaskExecuter.execute(SkipUpToDateTaskExecuter.java:59)
        at org.gradle.api.internal.tasks.execution.ResolveTaskOutputCachingStateExecuter.execute(ResolveTaskOutputCachingStateExecuter.java:54)
        at org.gradle.api.internal.tasks.execution.ValidatingTaskExecuter.execute(ValidatingTaskExecuter.java:59)
        at org.gradle.api.internal.tasks.execution.SkipEmptySourceFilesTaskExecuter.execute(SkipEmptySourceFilesTaskExecuter.java:101)
        at org.gradle.api.internal.tasks.execution.FinalizeInputFilePropertiesTaskExecuter.execute(FinalizeInputFilePropertiesTaskExecuter.java:44)
        at org.gradle.api.internal.tasks.execution.CleanupStaleOutputsExecuter.execute(CleanupStaleOutputsExecuter.java:91)
        at org.gradle.api.internal.tasks.execution.ResolveTaskArtifactStateTaskExecuter.execute(ResolveTaskArtifactStateTaskExecuter.java:62)
        at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:59)
        at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:54)
        at org.gradle.api.internal.tasks.execution.ExecuteAtMostOnceTaskExecuter.execute(ExecuteAtMostOnceTaskExecuter.java:43)
        at org.gradle.api.internal.tasks.execution.CatchExceptionTaskExecuter.execute(CatchExceptionTaskExecuter.java:34)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.run(EventFiringTaskExecuter.java:51)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:300)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:292)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:174)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:90)
        at org.gradle.internal.operations.DelegatingBuildOperationExecutor.run(DelegatingBuildOperationExecutor.java:31)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter.execute(EventFiringTaskExecuter.java:46)
        at org.gradle.execution.taskgraph.LocalTaskInfoExecutor.execute(LocalTaskInfoExecutor.java:42)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareWorkItemExecutor.execute(DefaultTaskExecutionGraph.java:277)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareWorkItemExecutor.execute(DefaultTaskExecutionGraph.java:262)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$ExecutorWorker$1.execute(DefaultTaskPlanExecutor.java:135)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$ExecutorWorker$1.execute(DefaultTaskPlanExecutor.java:130)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$ExecutorWorker.execute(DefaultTaskPlanExecutor.java:200)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$ExecutorWorker.executeWithWork(DefaultTaskPlanExecutor.java:191)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$ExecutorWorker.run(DefaultTaskPlanExecutor.java:130)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:63)
        at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:46)
        at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:55)
Caused by: java.lang.UnsatisfiedLinkError: Could not load J2V8 library. Reasons:
        no j2v8_macosx_x86_64 in java.library.path
        Native Library /Users/ntwigg/libj2v8_macosx_x86_64.dylib already loaded in another classloader

        at com.eclipsesource.v8.LibraryLoader.loadLibrary(LibraryLoader.java:75)
        at com.eclipsesource.v8.V8.load(V8.java:71)
        at com.eclipsesource.v8.V8.createV8Runtime(V8.java:145)
        ... 58 more
```

--- a/CHANGES.md
+++ b/CHANGES.md
@@ -10,6 +10,10 @@ This document is intended for Spotless developers.
 We adhere to the [keepachangelog](https://keepachangelog.com/en/1.0.0/) format (starting after version `1.27.0`).
 
 ## [Unreleased]
+### Added
+* `NodeJsGlobal.setSharedLibFolder` allows to set the location of nodejs shared libs. ([#586](https://github.com/diffplug/spotless/pull/586))
+### Fixed
+* Previously, the nodejs-based steps would throw `UnsatisfiedLinkError` if they were ever used from more than one classloader.  Now they can be used from any number of classloaders (important for gradle build daemon). ([#586](https://github.com/diffplug/spotless/pull/586))
 
 ## [1.31.0] - 2020-05-21
 ### Added

--- a/lib/src/main/java/com/diffplug/spotless/npm/NodeJSWrapper.java
+++ b/lib/src/main/java/com/diffplug/spotless/npm/NodeJSWrapper.java
@@ -21,7 +21,6 @@ import java.util.Map;
 import java.util.Objects;
 import java.util.Set;
 
-import com.diffplug.spotless.LineEnding;
 import com.diffplug.spotless.ThrowingEx;
 
 class NodeJSWrapper extends ReflectiveObjectWrapper {
@@ -36,9 +35,6 @@ class NodeJSWrapper extends ReflectiveObjectWrapper {
 	public NodeJSWrapper(ClassLoader classLoader) {
 		super(Reflective.withClassLoader(classLoader),
 				reflective -> {
-					if (LineEnding.PLATFORM_NATIVE.str().equals("\r\n")) {
-						reflective.invokeStaticMethod(V8_RUNTIME_CLASS, "setFlags", "-color=false"); // required to run prettier on windows
-					}
 					if (alreadySetup.add(classLoader)) {
 						// the bridge to node.js needs a .dll/.so/.dylib which gets loaded through System.load
 						// the problem is that when the JVM loads that DLL, it is bound to the specific classloader that called System.load

--- a/lib/src/main/java/com/diffplug/spotless/npm/NodeJSWrapper.java
+++ b/lib/src/main/java/com/diffplug/spotless/npm/NodeJSWrapper.java
@@ -16,11 +16,13 @@
 package com.diffplug.spotless.npm;
 
 import java.io.File;
+import java.util.HashSet;
 import java.util.Map;
 import java.util.Objects;
-import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.Set;
 
 import com.diffplug.spotless.LineEnding;
+import com.diffplug.spotless.ThrowingEx;
 
 class NodeJSWrapper extends ReflectiveObjectWrapper {
 
@@ -29,14 +31,40 @@ class NodeJSWrapper extends ReflectiveObjectWrapper {
 
 	public static final String WRAPPED_CLASS = "com.eclipsesource.v8.NodeJS";
 
-	private static final AtomicBoolean flagsSet = new AtomicBoolean(false);
+	private static final Set<ClassLoader> alreadySetup = new HashSet<>();
 
 	public NodeJSWrapper(ClassLoader classLoader) {
 		super(Reflective.withClassLoader(classLoader),
 				reflective -> {
-					final boolean firstRun = flagsSet.compareAndSet(false, true);
-					if (firstRun && LineEnding.PLATFORM_NATIVE.str().equals("\r\n")) {
+					if (LineEnding.PLATFORM_NATIVE.str().equals("\r\n")) {
 						reflective.invokeStaticMethod(V8_RUNTIME_CLASS, "setFlags", "-color=false"); // required to run prettier on windows
+					}
+					if (alreadySetup.add(classLoader)) {
+						// the bridge to node.js needs a .dll/.so/.dylib which gets loaded through System.load
+						// the problem is that when the JVM loads that DLL, it is bound to the specific classloader that called System.load
+						// no other classloaders have access to it, and if any other classloader tries to load it, you get an error
+						//
+						// ...but, you can copy that DLL as many times as you want, and each classloader can load its own copy of the DLL, and
+						// that is fine
+						//
+						// ...but the in order to do that, we have to manually load the DLL per classloader ourselves, which involves some
+						// especially hacky reflection into J2V8 *and* JVM internal
+						//
+						// so this is bad code, but it fixes our problem, and so far we don't have a better way...
+
+						// here we get the name of the DLL within the jar, and we get our own copy of it on disk
+						String resource = (String) reflective.invokeStaticMethodPrivate("com.eclipsesource.v8.LibraryLoader", "computeLibraryFullName");
+						File file = NodeJsGlobal.sharedLibs.nextDynamicLib(classLoader, resource);
+
+						// ideally, we would call System.load, but the JVM does some tricky stuff to
+						// figure out who actually called this, and it realizes it was Reflective, which lives
+						// outside the J2V8 classloader, so System.load doesn't work.  Soooo, we have to dig
+						// into JVM internals and manually tell it "this class from J2V8 called you"
+						Class<?> libraryLoaderClass = ThrowingEx.get(() -> classLoader.loadClass("com.eclipsesource.v8.LibraryLoader"));
+						reflective.invokeStaticMethodPrivate("java.lang.ClassLoader", "loadLibrary0", libraryLoaderClass, file);
+
+						// and now we set the flag in J2V8 which says "the DLL is loaded, don't load it again"
+						reflective.staticFieldPrivate("com.eclipsesource.v8.V8", "nativeLibraryLoaded", true);
 					}
 					return reflective.invokeStaticMethod(WRAPPED_CLASS, "createNodeJS");
 				});

--- a/lib/src/main/java/com/diffplug/spotless/npm/NodeJsGlobal.java
+++ b/lib/src/main/java/com/diffplug/spotless/npm/NodeJsGlobal.java
@@ -1,0 +1,93 @@
+/*
+ * Copyright 2016 DiffPlug
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.diffplug.spotless.npm;
+
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Optional;
+import java.util.stream.IntStream;
+
+import com.diffplug.spotless.LineEnding;
+import com.diffplug.spotless.ThrowingEx;
+
+/** Shared config acress the NodeJS steps. */
+public class NodeJsGlobal {
+	static SharedLibFolder sharedLibs;
+
+	static {
+		sharedLibs = new SharedLibFolder(
+				ThrowingEx.get(() -> Files.createTempDirectory("spotless-nodejs")));
+		sharedLibs.root.deleteOnExit();
+	}
+
+	/**
+	 * All of the NodeJS steps need to extract a bridge DLL for node.  By default this is
+	 * a random location, but you can set it to be anywhere.
+	 */
+	public static void setSharedLibFolder(File sharedLibFolder) {
+		sharedLibs = new SharedLibFolder(sharedLibFolder.toPath());
+	}
+
+	static class SharedLibFolder {
+		private final File root;
+
+		private SharedLibFolder(Path root) {
+			this.root = ThrowingEx.get(() -> root.toFile().getCanonicalFile());
+		}
+
+		static final int MAX_CLASSLOADERS_PER_CLEAN = 1_000;
+
+		synchronized File nextDynamicLib(ClassLoader loader, String resource) {
+			// find a new unique file
+			Optional<File> nextLibOpt = IntStream.range(0, MAX_CLASSLOADERS_PER_CLEAN)
+					.mapToObj(i -> new File(root, i + "_" + resource))
+					.filter(file -> !file.exists())
+					.findFirst();
+			if (!nextLibOpt.isPresent()) {
+				throw new IllegalArgumentException("Overflow, delete the spotless nodeJs cache: " + root);
+			}
+			File nextLib = nextLibOpt.get();
+			// copy the dll to it
+			try {
+				Files.createDirectories(nextLib.getParentFile().toPath());
+				try (FileOutputStream fileOut = new FileOutputStream(nextLib);
+						InputStream resourceIn = loader.loadClass("com.eclipsesource.v8.LibraryLoader").getResourceAsStream("/" + resource)) {
+					byte[] buf = new byte[0x1000];
+					while (true) {
+						int r = resourceIn.read(buf);
+						if (r == -1) {
+							break;
+						}
+						fileOut.write(buf, 0, r);
+					}
+				}
+			} catch (IOException | ClassNotFoundException e) {
+				throw ThrowingEx.asRuntime(e);
+			}
+			// make sure it is executable (on unix)
+			if (LineEnding.PLATFORM_NATIVE.str().equals("\n")) {
+				ThrowingEx.run(() -> {
+					Runtime.getRuntime().exec(new String[]{"chmod", "755", nextLib.getAbsolutePath()}).waitFor(); //$NON-NLS-1$
+				});
+			}
+			return nextLib;
+		}
+	}
+}

--- a/lib/src/main/java/com/diffplug/spotless/npm/NpmFormatterStepStateBase.java
+++ b/lib/src/main/java/com/diffplug/spotless/npm/NpmFormatterStepStateBase.java
@@ -103,7 +103,7 @@ abstract class NpmFormatterStepStateBase implements Serializable {
 		return new File(new File(this.nodeModulesDir, "node_modules"), this.npmConfig.getNpmModule());
 	}
 
-	private String j2v8MavenCoordinate() {
+	static String j2v8MavenCoordinate() {
 		return "com.eclipsesource.j2v8:j2v8_" + PlatformInfo.normalizedOSName() + "_" + PlatformInfo.normalizedArchName() + ":4.6.0";
 	}
 

--- a/lib/src/main/java/com/diffplug/spotless/npm/NpmFormatterStepStateBase.java
+++ b/lib/src/main/java/com/diffplug/spotless/npm/NpmFormatterStepStateBase.java
@@ -96,7 +96,7 @@ abstract class NpmFormatterStepStateBase implements Serializable {
 	}
 
 	protected NodeJSWrapper nodeJSWrapper() {
-		return new NodeJSWrapper(this.jarState.getClassLoader()); // TODO (simschla, 02.08.18): cache this instance
+		return new NodeJSWrapper(this.jarState.getClassLoader());
 	}
 
 	protected File nodeModulePath() {

--- a/lib/src/main/java/com/diffplug/spotless/npm/Reflective.java
+++ b/lib/src/main/java/com/diffplug/spotless/npm/Reflective.java
@@ -22,6 +22,10 @@ import java.util.Arrays;
 import java.util.Objects;
 import java.util.StringJoiner;
 
+import com.diffplug.spotless.ThrowingEx;
+
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+
 class Reflective {
 	private final ClassLoader classLoader;
 
@@ -53,6 +57,17 @@ class Reflective {
 	Object invokeStaticMethod(String className, String methodName, Object... parameters) {
 		try {
 			Method m = staticMethod(className, methodName, parameters);
+			return m.invoke(m.getDeclaringClass(), parameters);
+		} catch (IllegalAccessException | InvocationTargetException e) {
+			throw new ReflectiveException(e);
+		}
+	}
+
+	@SuppressFBWarnings("DP_DO_INSIDE_DO_PRIVILEGED")
+	Object invokeStaticMethodPrivate(String className, String methodName, Object... parameters) {
+		try {
+			Method m = staticMethod(className, methodName, parameters);
+			m.setAccessible(true);
 			return m.invoke(m.getDeclaringClass(), parameters);
 		} catch (IllegalAccessException | InvocationTargetException e) {
 			throw new ReflectiveException(e);
@@ -212,6 +227,18 @@ class Reflective {
 			return clazz.getDeclaredField(fieldName).get(clazz);
 		} catch (IllegalAccessException | NoSuchFieldException e) {
 			throw new ReflectiveException(e);
+		}
+	}
+
+	@SuppressFBWarnings("DP_DO_INSIDE_DO_PRIVILEGED")
+	void staticFieldPrivate(String className, String fieldName, boolean newValue) {
+		Class<?> clazz = clazz(className);
+		try {
+			Field field = clazz.getDeclaredField(fieldName);
+			field.setAccessible(true);
+			field.setBoolean(null, newValue);
+		} catch (NoSuchFieldException | SecurityException | IllegalArgumentException | IllegalAccessException e) {
+			throw ThrowingEx.asRuntime(e);
 		}
 	}
 

--- a/plugin-gradle/CHANGES.md
+++ b/plugin-gradle/CHANGES.md
@@ -7,6 +7,8 @@ We adhere to the [keepachangelog](https://keepachangelog.com/en/1.0.0/) format (
 * Support for ktfmt in KotlinGradleExtension. ([#583](https://github.com/diffplug/spotless/pull/583))
 ### Fixed
 * Users can now run `spotlessCheck` and `spotlessApply` in the same build. ([#584](https://github.com/diffplug/spotless/pull/584))
+* Fixed intermittent `UnsatisfiedLinkError` in nodejs-based steps. ([#586](https://github.com/diffplug/spotless/pull/586))
+  * Also, a shared library used by the nodejs steps used to be extracted into the user home directory, but now it is extracted into `{rootProject}/build/spotless-nodejs-cache`.
 
 ## [4.0.1] - 2020-05-21
 ### Fixed

--- a/plugin-gradle/src/main/java/com/diffplug/gradle/spotless/SpotlessExtension.java
+++ b/plugin-gradle/src/main/java/com/diffplug/gradle/spotless/SpotlessExtension.java
@@ -17,6 +17,7 @@ package com.diffplug.gradle.spotless;
 
 import static java.util.Objects.requireNonNull;
 
+import java.io.File;
 import java.lang.reflect.Constructor;
 import java.nio.charset.Charset;
 import java.nio.charset.StandardCharsets;
@@ -31,6 +32,7 @@ import org.gradle.api.plugins.BasePlugin;
 
 import com.diffplug.common.base.Errors;
 import com.diffplug.spotless.LineEnding;
+import com.diffplug.spotless.npm.NodeJsGlobal;
 
 public class SpotlessExtension {
 	final Project project;
@@ -63,6 +65,8 @@ public class SpotlessExtension {
 		if (registerDependenciesTask == null) {
 			registerDependenciesTask = project.getRootProject().getTasks().create(RegisterDependenciesTask.TASK_NAME, RegisterDependenciesTask.class);
 			registerDependenciesTask.setup();
+			// set where the nodejs runtime will put its temp dlls
+			NodeJsGlobal.setSharedLibFolder(new File(project.getBuildDir(), "spotless-nodejs-cache"));
 		}
 		this.registerDependenciesTask = registerDependenciesTask;
 	}

--- a/plugin-maven/CHANGES.md
+++ b/plugin-maven/CHANGES.md
@@ -3,6 +3,8 @@
 We adhere to the [keepachangelog](https://keepachangelog.com/en/1.0.0/) format (starting after version `1.27.0`).
 
 ## [Unreleased]
+### Fixed
+* Shared library used by the nodejs-based steps used to be extracted into the user home directory, but now it is extracted into a temporary directory and deleted on VM shutdown. ([#586](https://github.com/diffplug/spotless/pull/586))
 
 ## [1.31.1] - 2020-05-21
 ### Fixed

--- a/testlib/src/test/java/com/diffplug/spotless/npm/NodeJsNativeDoubleLoadTest.java
+++ b/testlib/src/test/java/com/diffplug/spotless/npm/NodeJsNativeDoubleLoadTest.java
@@ -1,0 +1,53 @@
+/*
+ * Copyright 2016 DiffPlug
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.diffplug.spotless.npm;
+
+import java.util.Optional;
+
+import org.assertj.core.api.Assertions;
+import org.junit.Test;
+
+import com.diffplug.common.collect.ImmutableMap;
+import com.diffplug.spotless.JarState;
+import com.diffplug.spotless.TestProvisioner;
+
+public class NodeJsNativeDoubleLoadTest {
+	@Test
+	public void inMultipleClassLoaders() throws Exception {
+		JarState state = JarState.from(NpmFormatterStepStateBase.j2v8MavenCoordinate(), TestProvisioner.mavenCentral());
+		ClassLoader loader1 = state.getClassLoader(1);
+		ClassLoader loader2 = state.getClassLoader(2);
+		createAndTestWrapper(loader1);
+		createAndTestWrapper(loader2);
+	}
+
+	@Test
+	public void multipleTimesInOneClassLoader() throws Exception {
+		JarState state = JarState.from(NpmFormatterStepStateBase.j2v8MavenCoordinate(), TestProvisioner.mavenCentral());
+		ClassLoader loader3 = state.getClassLoader(3);
+		createAndTestWrapper(loader3);
+		createAndTestWrapper(loader3);
+	}
+
+	private void createAndTestWrapper(ClassLoader loader) throws Exception {
+		try (NodeJSWrapper node = new NodeJSWrapper(loader)) {
+			V8ObjectWrapper object = node.createNewObject(ImmutableMap.of("a", 1));
+			Optional<Integer> value = object.getOptionalInteger("a");
+			Assertions.assertThat(value).hasValue(1);
+			object.release();
+		}
+	}
+}

